examples/iso_dh/DY.Example.DH.Protocol.Total.fst

```diff
@@ -0,0 +1,167 @@
+
+%splice [ps_message] (gen_parser (`message))
+%splice [ps_message_is_well_formed] (gen_is_well_formed_lemma (`message))
+
+instance parseable_serializeable_message: parseable_serializeable bytes message = mk_parseable_serializeable ps_message
+
+// Definition of signature terms
+[@@ with_bytes bytes]
+type sig_message2 = {
+  alice:principal;
+  gx:bytes;
+  gy:bytes;
+}
+
+%splice [ps_sig_message2] (gen_parser (`sig_message2))
+%splice [ps_sig_message2_is_well_formed] (gen_is_well_formed_lemma (`sig_message2))
+
+[@@ with_bytes bytes]
+type sig_message3 = {
+  bob:principal;
+  gx:bytes;
+  gy:bytes;
+}
+
+%splice [ps_sig_message3] (gen_parser (`sig_message3))
+%splice [ps_sig_message3_is_well_formed] (gen_is_well_formed_lemma (`sig_message3))
+
+[@@ with_bytes bytes]
+type sig_message =
+  | SigMsg2: msg:sig_message2 -> sig_message
+  | SigMsg3: msg:sig_message3 -> sig_message
+
+%splice [ps_sig_message] (gen_parser (`sig_message))
+%splice [ps_sig_message_is_well_formed] (gen_is_well_formed_lemma (`sig_message))
+
+instance parseable_serializeable_sig_message: parseable_serializeable bytes sig_message = mk_parseable_serializeable ps_sig_message
+
+
+(*** Message Processing ***)
+
+// Alice generates message 1
+val compute_message1: principal -> bytes -> bytes
+let compute_message1 alice x =
+  let gx = dh_pk x in
+  let msg = Msg1 {alice; gx} in
+  serialize message msg
+
+// Bob parses message 1
+val decode_message1: bytes -> option message1
+let decode_message1 msg1_bytes =
+  let? msg1 = parse message msg1_bytes in
+  // These lines are the...
+  guard (Msg1? msg1);?
+  Some (Msg1?.msg msg1)
+  // ...short version of the following match:
+  (*    
+  match msg1 with
+  | Msg1 msg -> Some msg
+  | _ -> None
+  *)
+
+// Bob generates message 2
+val compute_message2: principal -> principal -> bytes -> bytes -> bytes -> bytes -> bytes
+let compute_message2 alice bob gx gy sk_b n_sig =
+  let sig_msg = SigMsg2 {alice; gx; gy} in
+  let sg = sign sk_b n_sig (serialize sig_message sig_msg) in
+  let msg = Msg2 {bob; gy; sg} in
+  serialize message msg
+
+// Alice parses message 2
+type verify_msg2_result = {sg:bytes; gy:bytes; gx:bytes; k:bytes}
+val decode_and_verify_message2: bytes -> principal -> bytes -> bytes -> option verify_msg2_result
+let decode_and_verify_message2 msg2_bytes alice x pk_b =
+  let? msg2_parsed = parse message msg2_bytes in
+  guard (Msg2? msg2_parsed);?
+  let msg2 = Msg2?.msg msg2_parsed in
+  // Verify the signature contained in the message 2
+  // with the gy value from the message and the gx
+  // value from Alice's state.
+  let gx = dh_pk x in
+  let gy = msg2.gy in
+  let sig_msg = SigMsg2 {alice; gx; gy} in
+  // These lines are the...
+  guard(verify pk_b (serialize sig_message sig_msg) msg2.sg);?
+  let k = dh x gy in
+  Some {sg=msg2.sg; gy; gx; k}
+  // ...short version of the following if-else block:
+  (*  
+  if verify pk_b (serialize sig_message sig_msg) msg2.sg then Some {msg2; gx; k}
```

Comment:
The comment do not correspond to the code above

Reply:
So, should I use the if-else block?